### PR TITLE
Feature : requires key

### PR DESCRIPTION
**Pull request by sdrdis**
This pull request adds support for the `requires` key in the metadata. Here are precise specifications:
- value can be a string or an array. If it is a string, it will be considered as a array with a single element.
- Application can now require other application. This has two implications :
  - When the user installs an application, the core will try to install all required applications. If some of these applications don't exist, installation is judged impossible.
  - It is impossible to uninstall an application if other application requires it.
- If the `requires` key is not defined but the `extend` key is, the core considers that the application requires the one it is extending.
- The application manager interface has been improved to manage this `requires` key

### CODE REVIEW

File: framework/applications/noviusos_appmanager/views/admin/app_manager.view.php [hunk 1]
```diff
@@ -67,7 +67,33 @@
...
+    <?php
+    if ($app->canUninstall()) {
+        ?>
                             <a href="#" data-app="<?= htmlspecialchars(\Format::forge(array('name' => $app->folder, 'action' => 'remove'))->to_json()) ?>" onclick="return false;"><button data-icon="arrowthick-1-s"><?= __('Uninstall') ?></button></a>
+    <?php
+    } else {
+        $dependents = $app->installedDependentApplications();
+        foreach ($dependents as &$dependent) {
+            $dependent = \Nos\Application::forge($dependent)->get_name();
+        }
```
Comment (Savageman) [inline framework/applications/noviusos_appmanager/views/admin/app_manager.view.php:82]: Don't forget to unset($dependent) to free the reference.

File: framework/classes/application.php [hunk 2]
```diff
@@ -289,6 +390,11 @@ public function install($add_permission = true)
      */
     public function uninstall()
     {
+        if (!$this->canUninstall()) {
+            $dependents = $this->installedDependentApplications();
+            throw new \Exception('Application '.$this->folder.
+                ' can\'t be uninstalled because it needs the following applications: '.implode(', ', $dependents).'.');
```
Comment (Savageman) [inline framework/classes/application.php]: Wrong label? because it needs -> because it's needed by?

File: framework/classes/application.php [hunk 3]
```diff
@@ -10,10 +10,13 @@
 
 namespace Nos;
 
+use Oil\Exception;
```
Comment (Savageman) [inline framework/classes/application.php]: Is it used? Can't see where.

File: framework/applications/noviusos_appmanager/views/admin/app_manager.view.php (hunk 1 above)
Comment (sdrdis) [inline framework/applications/noviusos_appmanager/views/admin/app_manager.view.php:82]: good catch :)

File: framework/classes/application.php (hunk 2 above)
Comment (sdrdis) [inline framework/classes/application.php]: yep :)

File: framework/classes/application.php (hunk 3 above)
Comment (sdrdis) [inline framework/classes/application.php]: IDE automation...

**Comment by sdrdis** [comment]
I will merge this afternoon if there aren't any more suggestions...

**Comment by felixgilles** [comment]
It's good for me

**Pull request merged**